### DISCUSSION TITLE
Automatic Different for Hydro-mechanical simulation

### DISCUSSION THREAD

**Discussion by alexchen4ai** [idaholab/moose#19188] [Q&A General] [answered]
Hi Moose experts,

I plan to use moose to implement the basic hydro-mechanical simulation.

I haven't tried to use Automatic differentiation for porous flow before. The equations I want to implement is straightforward. 
![image](https://user-images.githubusercontent.com/52059011/138575746-aa9278c6-a2ab-49d6-8b44-553917c84466.png)
Here, \omega is darcy's velocity. 

After I finish the implementation using AD, I can't get the converged result.
` 0 Nonlinear |R| = 9.839972e-02
  Linear solve did not converge due to DIVERGED_NANORINF iterations 0
Nonlinear solve did not converge due to DIVERGED_LINEAR_SOLVE iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge`

I want to ask if we can use AD in porous flow problem? I am currently using PJFNK solver. Is it appropriate for this simulation? Thank you!

**Comment by GiudGiud** [comment]
Hello

If you are using AD you can use Newton instead of PJFNK since the Jacobian you're creating should be exact. This is also the recommended option for porous flow. 

But that might not be the issue. 
This very first nonlinear residual, is it the same as with a non-AD implementation? It should.
Is it possible to only solve one of the two equations, keeping all other terms constant, for this application? To verify the correctness of the kernels

Did you look at this porous flow documentation https://mooseframework.inl.gov/modules/porous_flow/tutorial_02.html
has information on numerical solves

Guillaume

**Reply by alexchen4ai** [reply]
Thank you! Since I plan to include some new constituive laws which are pretty complicated, I want to use AD to compute the jacobian automatically. And this simple equation will be my starting point to include more complicated stuff. 

I can solve the two equations seperately. But when I add the residual term of `(_strain[_qp].trace() - _strain_old[_qp].trace())/_dt*_test[_i][_qp];` for the second equations. The result doesn't converge. So, to deal with porous flow problem, we should implement the jacobian exactly, right? I tried the methods in tutorial but it doesn't work here. Now, the running issues become
`Time Step 1, time = 0.1, dt = 0.1

Performing automatic scaling calculation

 0 Nonlinear |R| = 6.380534e+00
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 30
 1 Nonlinear |R| = 5.892391e+00
  Linear solve did not converge due to DIVERGED_BREAKDOWN iterations 30
 2 Nonlinear |R| = 5.361072e+00
  Linear solve did not converge due to DIVERGED_ITS iterations 300
 3 Nonlinear |R| = 4.711670e+00
  Linear solve did not converge due to DIVERGED_ITS iterations 300
 4 Nonlinear |R| = 4.711670e+00
  Linear solve did not converge due to DIVERGED_ITS iterations 300
 5 Nonlinear |R| = 4.711670e+00
  Linear solve did not converge due to DIVERGED_ITS iterations 300
 6 Nonlinear |R| = 4.711670e+00
  Linear solve did not converge due to DIVERGED_ITS iterations 300
 7 Nonlinear |R| = 4.711670e+00
  Linear solve did not converge due to DIVERGED_ITS iterations 300
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 7
 Solve Did NOT Converge!
Aborting as solve did not converge`

**Reply by GiudGiud** [reply]
There's something special about implementing time derivative kernels. If you do it like this you wont be able to leverage all the cool time steppers we have. You have to use the dot functions like the Time Kernels. Unfortunately I m not sure where good documentation for doing this is. @lindsayad do you know?

What is _strain? To check if this type has a dot() function

Maybe have a look at the AD Time Derivative kernel, https://github.com/idaholab/moose/blob/next/framework/src/kernels/ADTimeDerivative.C
which inherits from the TimeKernel, which defines u_dot https://github.com/idaholab/moose/blob/next/framework/src/kernels/ADTimeKernel.C

Other page of interest:
https://mooseframework.inl.gov/source/executioners/Transient.html

**Reply by lindsayad** [reply]
The first thing I would do is compile in dbg mode, like `METHOD=dbg make -j8` and then run with the `dbg` executable and see if you have any floating point exceptions or things like that. I suspect that you do given the errors out of your linear solver.

**Reply by lindsayad** [reply]
More documentation on debugging is [here](https://mooseframework.inl.gov/application_development/debugging.html)

**Reply by alexchen4ai** [reply]
Thank you. I tried both methods you mentioned. I guess the main issue is that the permeability I choose is too low.  To solve this kind of problem, it is better to use explicit jacobian expression.

**Reply by GiudGiud** [reply]
I'm not an expert on porous flow so I cant comment on that. As a general rule though, AD will give you a better Jacobian than explicit. 
@WilkAndy may know what s going on

**Reply by alexchen4ai** [reply]
> I'm not an expert on porous flow so I cant comment on that. As a general rule though, AD will give you a better Jacobian than explicit. @WilkAndy may know what s going on

Thank you. Yeah, I can get the code to run right now for a large permeability I choose. Note that permeability is used in darcy's law
$$\vec\omega = \frac{\kappa}{\mu}\nabla p$$

However, I can't get the converged result if kappa is small.

**Reply by hugary1995** [reply]
I solve these equations routinely with AD, it for sure works. I second @lindsayad 's suggestion -- it's likely you have a floating point exception somewhere, probably in your constitutive model.

**Comment by WilkAndy** [comment, marked as answer]
From the above discussion, it is difficult to know whether you're using the `PorousFlow` module, or your own code.  If you're using the `PorousFlow` module (and i recommend you do, because this is a simple scenario for `PorousFlow`) then this physics should be easily modelled using [PorousFlowFullySaturated](https://mooseframework.inl.gov/source/actions/PorousFlowFullySaturated.html).  Please also read the [tutorial pages](https://mooseframework.inl.gov/modules/porous_flow/tutorial_00.html).  You probably should never use automatic differentiation in `PorousFlow` models, since all the Jacobian entries are already coded, and you should almost always use `solve_type = Newton` (i don't think any of the `PorousFlow` tests or examples use anything else, because in more complicated `PorousFlow` situations, other nonlinear solvers take too long to converge).

**Reply by alexchen4ai** [reply]
> From the above discussion, it is difficult to know whether you're using the `PorousFlow` module, or your own code. If you're using the `PorousFlow` module (and i recommend you do, because this is a simple scenario for `PorousFlow`) then this physics should be easily modelled using [PorousFlowFullySaturated](https://mooseframework.inl.gov/source/actions/PorousFlowFullySaturated.html). Please also read the [tutorial pages](https://mooseframework.inl.gov/modules/porous_flow/tutorial_00.html). You probably should never use automatic differentiation in `PorousFlow` models, since all the Jacobian entries are already coded, and you should almost always use `solve_type = Newton` (i don't think any of the `PorousFlow` tests or examples use anything else, because in more complicated `PorousFlow` situations, other nonlinear solvers take too long to converge).

Hi Wilk, I am using my own code. The`PorousFlow` module has explicit jacobian and I used it before. Now, I am considering to implement double porosity coupled with plasticity in Moose. But the jacobian is expected to be complex. So, I consider to use AD to compute the jacobian automatically and accurately. I try to use AD to solve the equations above since I want to use it as my starting point for some really complex equations and test if AD could be used for porous flow problem.

**Reply by WilkAndy** [reply]
Aha, i get it.   Sounds fun!   Probably you need some sort of `scaling` on your Variables.  Even if you aren't using `PorousFlow`, [this page](https://mooseframework.inl.gov/modules/porous_flow/convergence.html) might be useful to you.

**Reply by alexchen4ai** [reply]
> Aha, i get it. Sounds fun! Probably you need some sort of `scaling` on your Variables. Even if you aren't using `PorousFlow`, [this page](https://mooseframework.inl.gov/modules/porous_flow/convergence.html) might be useful to you.

Thank you for the sharing!

**Reply by zhang-rite** [reply]
> You probably should never use automatic differentiation in `PorousFlow` models

@WilkAndy, it looks like AD can give a better Jacobian than explicit as @GiudGiud wrote. Why does `PorousFlow` module not?

**Reply by lindsayad** [reply]
If a developer can take the time and write out an exact hand Jacobian, the simulation will run roughly 1.5 times faster

**Reply by WilkAndy** [reply]
> > You probably should never use automatic differentiation in `PorousFlow` models
> 
> @WilkAndy, it looks like AD can give a better Jacobian than explicit as @GiudGiud wrote. Why does `PorousFlow` module not?

@zhang-rite - do you have evidence of this in `PorousFlow` (if you do, we need to urgently investigate why the hand-coded derivatives are incorrect since that could signal a major problem)?  If not, then: (1) the hand-coded Jacobian is faster as @lindsayad wrote, and (2) AD wasn't around when PorousFlow was written, so we didn't have that option, boohoo.

**Reply by zhang-rite** [reply]
@WilkAndy @lindsayad Thank you for the reply. I get to understand the situation. Before seeing this discussion, I just read the recent paper "[Lindsay, Alexander, et al. "Automatic Differentiation in MetaPhysicL and Its Applications in MOOSE." Nuclear Technology (2021): 1-18](https://doi.org/10.1080/00295450.2020.1838877)" and think AD is so attractive for "the guarantee of an exact Jacobian at a relatively small overhead cost". 
Anther open-source code MRST (MATLAB based, desktop only) for reservoir modeling seems beneficial in speed from the usage AD.
